### PR TITLE
Filtering subscribes in `ActionEvaluationPublisher`

### PR DESCRIPTION
**Pull request by boscohyun**
- Filtering subscribes in `ActionEvaluationPublisher`

### CODE REVIEW

File: NineChronicles.Standalone/ActionEvaluationPublisher.cs [hunk 1]
```diff
@@ -149,5 +163,58 @@ public override async Task StopAsync(CancellationToken cancellationToken)
...
+            return false;
+        }
+
+        private void ResetAddressesToSubscribe(ActionBase.ActionEvaluation<ActionBase> ev)
+        {
+            if (!(ev.Action is CreateAvatar _) &&
+                !(ev.Action is CreateAvatar2 _))
+            {
+                return;
+            }
```
Comment (ipdae) [inline NineChronicles.Standalone/ActionEvaluationPublisher.cs:194]: What is the reason for excluding CreateAvatar?

Comment (boscohyun) [reply NineChronicles.Standalone/ActionEvaluationPublisher.cs:194]: @ipdae This line is including `CreateAvatar` and `CreateAvatar2` action. When if not both and return.

**Review by limebell — APPROVED** [review]

**Review by riemannulus — APPROVED** [review]

**Review by Unengine — APPROVED** [review]

File: NineChronicles.Standalone/Controllers/GraphQLController.cs [hunk 2]
```diff
@@ -92,6 +92,17 @@ public IActionResult SetPrivateKey([FromBody] SetPrivateKeyRequest request)
...
 
+            var agentAddress = privateKey.ToAddress();
+            // FIXME: `StandaloneContext.BlockChain` can be 'null'
+            var chainAgentState = StandaloneContext.BlockChain?.GetState(agentAddress);
+            if (chainAgentState != null)
+            {
+                var agentState = new AgentState((Bencodex.Types.Dictionary) chainAgentState);
+                StandaloneContext.NineChroniclesNodeService.ExceptionRenderer.RenderAgentAndAvatarAddresses(
+                    agentAddress,
+                    agentState.avatarAddresses.Values.ToList());
```
Comment (limebell) [inline NineChronicles.Standalone/Controllers/GraphQLController.cs:103]: Can `NullReferenceException` occur in this statement?

Comment (boscohyun) [reply NineChronicles.Standalone/Controllers/GraphQLController.cs:103]: No. Assign the `avatarAddresses` in the `AgentState` constructor.

**Review by limebell — APPROVED** [review]

File: NineChronicles.Standalone/ActionEvaluationPublisher.cs [hunk 3]
```diff
@@ -167,5 +182,65 @@ public override async Task StopAsync(CancellationToken cancellationToken)
...
+            {
+                return false;
+            }
+
+            if (ev.OutputStates.UpdatedFungibleAssets.ContainsKey(_agentAddress))
+            {
+                return true;
+            }
+            
+            var updatedAddresses = ev.OutputStates.UpdatedAddresses;
```
Comment (ipdae) [inline NineChronicles.Standalone/ActionEvaluationPublisher.cs:199]: i think this is missing update state when throw exception in IAction.Execute

File: NineChronicles.Standalone/ActionEvaluationPublisher.cs [hunk 4]
```diff
@@ -167,5 +182,66 @@ public override async Task StopAsync(CancellationToken cancellationToken)
...
+            
+            var agentState = new AgentState((Bencodex.Types.Dictionary) chainAgentState);
+            ResetAddressesToSubscribe((_agentAddress, agentState.avatarAddresses.Values.ToList()));
+        }
+        
+        private void ResetAddressesToSubscribe((Address agentAddress, List<Address> avatarAddresses) tuple)
+        {
+            Log.Debug($"ResetAddressesToSubscribe() invoked. {tuple.agentAddress} {tuple.avatarAddresses?.Count ?? 0}");
+            _agentAddress = tuple.agentAddress;
+            _addressesToSubscribe = new List<Address> {_agentAddress};
```
Comment (ipdae) [inline NineChronicles.Standalone/ActionEvaluationPublisher.cs:239]: i think we can just subscribe `_agentAddress`.

File: NineChronicles.Standalone/ActionEvaluationPublisher.cs [hunk 5]
```diff
@@ -167,5 +182,66 @@ public override async Task StopAsync(CancellationToken cancellationToken)
...
+                !_addressesToSubscribe.Any())
+            {
+                return false;
+            }
+
+            if (ev.Signer.Equals(_agentAddress) ||
+                ev.OutputStates.UpdatedFungibleAssets.ContainsKey(_agentAddress))
+            {
+                return true;
+            }
```
Comment (ipdae) [inline NineChronicles.Standalone/ActionEvaluationPublisher.cs:198]:
```suggestion
            return (ev.Signer.Equals(_agentAddress) || ev.OutputStates.UpdatedFungibleAssets.ContainsKey(_agentAddress))
```
```

File: NineChronicles.Standalone/ActionEvaluationPublisher.cs (hunk 4 above)
Comment (boscohyun) [reply NineChronicles.Standalone/ActionEvaluationPublisher.cs:239]: I will leave only `_agentAddress` through next PR. @ipdae

**Review by riemannulus — APPROVED** [review]

**Pull request merged**